### PR TITLE
Scaffold initial directory structure

### PR DESCRIPTION
**Pull request by CjS77**
From the README:

The code follows a domain-driven design layout, with top-level folders falling into infrastructure, domain, or application layers.

 The `infrastructure` folder contains code that is not Tari-specific. It holds the following crates:

* `comms`: The networking and messaging subsystem
* `crypto`: All cryptographic services, including a Curve25519 implementation
* `storage`: Data persistence services, including LMDB
 The `base_layer` is a domain-level folder and contains:
* `core`: common classes and traits, such as `Transaction`s and `Block`s
* `blockchain`: The Tari consensus code
* `mempool`: The unconfirmed transaction pool implementation
* `mining`: The merge-mining modules
* `p2p`: The block and transaction propagation module
* `api`: interfaces for clients and wallets to interact with the base layer components

 The `digital_assets_layer` is a domain-level folder contains code related to the management of native Tari digital assets. Substructure TBD.

 It's envisaged that at least the following `applications` are built on top of the domain layer libraries:

* A standalone miner (`tari_miner`)
* A pool miner (`tari_pool_miner`)
* A CLI wallet for the Tari cryptocurrency (`cli_wallet`)
* A full node executable (`tari_basenode`)

### CODE REVIEW

**Comment by CjS77** [comment]
@philipr-za thanks for the catch. git doesn't allow empty folders, so I've added Cargo scaffolding to the `infrastructure` folders and repushed

File: Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+[workspace]
+
+members = [
+    "base_layer/blockchain",
```
Comment (mikethetike) [inline Cargo.toml]: For some reason, most chains I've seen call this folder `chain`

File: Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    "base_layer/core",
```
Comment (mikethetike) [inline Cargo.toml]: I would have put all the `base_layer` folders in the root, and just had the `digital_assets_layer` as `digital_assets`

File: Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    "base_layer/mempool",
```
Comment (mikethetike) [inline Cargo.toml]: Time will tell whether we need a `mempool` folder. I would think that most of this will be in `chain` or `core`

**Comment by neonknight64** [comment]
It looks good. Where will the unit and functional tests fit in?

File: Cargo.toml (hunk 2 above)
Comment (CjS77) [reply Cargo.toml]: I've been back and forth on this myself, but ultimately came to the conclusion that the base and DA layers are siblings and should sit at the same level.

There is scope for a "common" crate that holds stuff that everyone will use, like error handling, logging, etc.

File: Cargo.toml (hunk 3 above)
Comment (CjS77) [reply Cargo.toml]: Sure.

**Comment by CjS77** [comment]
Just follow Rust conventions, Unit tests are in the source code; functional tests will be in a `tests` folder per crate etc.

File: Cargo.toml (hunk 1 above)
Comment (CjS77) [reply Cargo.toml]: Personally, I prefer the more explicit `blockchain`, but not married to it

**Comment by hansieodendaal** [comment]
It is a great start. Something I immediately observed to be missing, the consensus algorithms/schemes/protocols. Where would they live? In `infrastructure` ? This has potential use in base layer and 2nd layer.

**Comment by CjS77** [comment]
Possibly `crypto`; If it gets too cumbersome, e.g. we have a fairly involved DAG implementation, we can pull it out into it's own crate

**Comment by CjS77** [comment]
Yeah, it's just a start; and we can refactor obviously; but this type of thing tends to be sticky so worth having the convo about it

**Pull request merged**